File: SearchEngine/InvertedIndex.js
class InvertedIndex {
  constructor() {
    this.index = Object.create(null);
    this.docCount = 0;
  }

  addDocument(docId, tokens) {
    this.docCount++;

    const freqMap = new Map();
    for (const token of tokens) {
      freqMap.set(token, (freqMap.get(token) || 0) + 1);
    }

    for (const [token, count] of freqMap.entries()) {
      if (!this.index[token]) {
        this.index[token] = { df: 0, docs: {} };
      }
      this.index[token].df++;
      this.index[token].docs[docId] = count;
    }
  }

  getPosting(token) {
    return this.index[token] || null;
  }
}

module.exports = { InvertedIndex };

File: Go/SearchEngine-go/inverted_index.go
package main

type Posting struct {
	DF   int
	Docs map[string]int
}

type InvertedIndex struct {
	Index    map[string]*Posting
	DocCount int
}

func NewInvertedIndex() *InvertedIndex {
	return &InvertedIndex{Index: make(map[string]*Posting)}
}

func (i *InvertedIndex) AddDocument(docID string, tokens []string) {
	i.DocCount++
	freq := make(map[string]int)
	for _, token := range tokens {
		freq[token]++
	}
	for token, count := range freq {
		if i.Index[token] == nil {
			i.Index[token] = &Posting{Docs: make(map[string]int)}
		}
		i.Index[token].DF++
		i.Index[token].Docs[docID] = count
	}
}

func (i *InvertedIndex) GetPosting(token string) *Posting {
	return i.Index[token]
}
